plugins/mediatek-scaler/fu-mediatek-scaler-device.c

```diff
@@ -0,0 +1,849 @@
+	report_data_sz = buf[1] - DDC_DATA_LEN_DFT;
+	if (report_data_sz + 3 > sizeof(buf)) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "invalid read buffer: size 0x%02x exceeded 0x%02lx.",
+			    (guint)report_data_sz,
+			    sizeof(buf));
+		return FALSE;
+	}
+
+	/* verify read buffer: size plus 2 to incorporate opcode and vcp code */
+	checksum ^= DDC_HOST_I2C_ADDR_CHKSUM;
+	for (gsize i = 0; i < report_data_sz + 2; i++)
+		checksum ^= buf[i];
+
+	if (buf[report_data_sz + 2] != checksum) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "invalid read buffer, checksum expected 0x%02x, got 0x%02x.",
+			    checksum,
+			    buf[report_data_sz + 2]);
+		return FALSE;
+	}
+
+	/* checksum is needn't */
+	g_byte_array_append(readbuf, buf, report_data_sz + 2);
+
+	/* print the raw data */
+	rawbytes = g_bytes_new(readbuf->data, (gsize)readbuf->len);
+	fu_dump_bytes(G_LOG_DOMAIN, "DDC/CI read report", rawbytes);
+	return TRUE;
+}
+
+static gboolean
+fu_mediatek_scaler_device_set_ddc_priority(FuDevice *device, guint8 flag, GError **error)
+{
+	FuMediatekScalerDevice *self = FU_MEDIATEK_SCALER_DEVICE(device);
+	g_autoptr(GByteArray) cod = g_byte_array_new();
+	g_autoptr(GError) error_local = NULL;
+	const guint8 buf[2] = {DDC_OPCODE_MTK_REQ, DDC_VCP_CODE_PRIORITY};
+
+	g_byte_array_append(cod, buf, sizeof(buf));
+	g_byte_array_append(cod, &flag, sizeof(guint8));
+	if (!(mediatek_scaler_device_ddc_write(self, cod, &error_local))) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "failed to set priority '0x%u', unsupported display? %s",
+			    flag,
+			    (error_local != NULL ? error_local->message : "unknown"));
+		fu_device_sleep(device, 10);
+		return FALSE;
+	}
+	fu_device_sleep(device, 10);
+	return TRUE;
+}
+
+static gboolean
+fu_mediatek_scaler_display_is_connected(FuDevice *device, GError **error)
+{
+	FuMediatekScalerDevice *self = FU_MEDIATEK_SCALER_DEVICE(device);
+	g_autoptr(GByteArray) cmdbuf = g_byte_array_new();
+	g_autoptr(GByteArray) readbuf = g_byte_array_new();
+	g_autoptr(GError) error_local = NULL;
+	guint8 randval1 = g_random_int_range(1, 255);
+	guint8 randval2 = g_random_int_range(1, 255);
+	const guint8 cmd[] = {DDC_OPCODE_MTK_REQ, DDC_VCP_CODE_SUM, randval1, randval2};
+
+	g_byte_array_append(cmdbuf, cmd, sizeof(cmd));
+	if (mediatek_scaler_device_ddc_read(self, cmdbuf, readbuf, &error_local)) {
+		/* device unique feature */
+		if (readbuf->data[3] == (randval1 + randval2)) {
+			g_info("found mediatek display controller: %s, i2c-dev: %s",
+			       self->dp_aux_dev_name,
+			       self->i2c_dev_name);
+			return TRUE;
+		}
+	}
+	g_set_error(error,
+		    FWUPD_ERROR,
+		    FWUPD_ERROR_NOT_SUPPORTED,
+		    "unsuccessful feature test, could be unsupported display: %s",
+		    (error_local != NULL ? error_local->message : "unknown"));
+	return FALSE;
```

Comment:
I don't see a reason to use error_local in this function.  You can simplify a lot of this by just using error in `mediatek_scaler_device_ddc_read` and then returning false if it fails.

Reply:
if `mediatek_scaler_device_ddc_read` fails then IO error introduces probe() failure. The `error_local` filters unsupported dispaly intentionally. how does that sound to you

Reply:
Then I think you should be using g_steal_pointer in the fail case fo promote the error rather than steal the string and make a new one.

Reply:
This should have been fixed though?